BaseInterfaces/src/array.jl

```diff
@@ -0,0 +1,93 @@
+#= Abstract Arrays
+
+Methods to implement		            Brief description
+size(A)		                            Returns a tuple containing the dimensions of A
+getindex(A, i::Int)		                (if IndexLinear) Linear scalar indexing
+getindex(A, I::Vararg{Int, N})          (if IndexCartesian, where N = ndims(A)) N-dimensional scalar indexing
+
+Optional methods	                    Default definition	                    Brief description
+IndexStyle(::Type)	                    IndexCartesian()	                    Returns either IndexLinear() or IndexCartesian(). See the description below.
+setindex!(A, v, i::Int)		                                                    (if IndexLinear) Scalar indexed assignment
+setindex!(A, v, I::Vararg{Int, N})                                              (if IndexCartesian, where N = ndims(A)) N-dimensional scalar indexed assignment
+getindex(A, I...)	                    defined in terms of scalar getindex	    Multidimensional and nonscalar indexing
+setindex!(A, X, I...)	                defined in terms of scalar setindex!	Multidimensional and nonscalar indexed assignment 
+iterate	                                defined in terms of scalar getindex	Iteration
+length(A)	                            prod(size(A))	                        Number of elements
+similar(A)	                            similar(A, eltype(A), size(A))	        Return a mutable array with the same shape and element type
+similar(A, ::Type{S})	                similar(A, S, size(A))	                Return a mutable array with the same shape and the specified element type
+similar(A, dims::Dims)	                similar(A, eltype(A), dims)	            Return a mutable array with the same element type and size dims
+similar(A, ::Type{S}, dims::Dims)	    Array{S}(undef, dims)	                Return a mutable array with the specified element type and size
+
+Non-traditional indices	                Default definition	                    Brief description
+axes(A)	                                map(OneTo, size(A))	                    Return a tuple of AbstractUnitRange{<:Integer} of valid indices
+similar(A, ::Type{S}, inds)	            similar(A, S, Base.to_shape(inds))	    Return a mutable array with the specified indices inds (see below)
+similar(T::Union{Type,Function}, inds)	T(Base.to_shape(inds))	                Return an array similar to T with the specified indices inds (see below)
+=#
+
+# And arbitrary new type for array values
+struct ArrayTestVal
+    a::Int
+end
+
+# In case `eltype` and `ndims` have custom methods
+# We should always be able to use these to mean the same thing
+_eltype(::AbstractArray{T}) where T = T
+_ndims(::AbstractArray{<:Any,N}) where N = N
+
+array_components = (;
+    mandatory = (;
+        type = A -> A isa AbstractArray,
+        eltype = (
+            A -> eltype(A) isa Type,
+            A -> eltype(A) == _eltype(A),
+        ),
+        ndims = (
```

Comment:
This should have a fallback definition already, no? IIRC `ndims` comes from having `size` defined already.

Reply:
Doesnt it use the second type parameter of AbstractArray? 

I just wanted to check that was not messed up in the implementation of a custom type. Maybe not titally useful

Reply:
True, but that's not an issue because you have to declare where in a subtype the second parameter of a parent type goes. E.g. it doesn't matter that I can swap the parameters around in my definition:

```
julia> struct MyArr{N,T} <: AbstractArray{T,N} end

julia> ndims(MyArr{2,Int})
2
```